cpu/[msp|cc]430: added peripheral timer driver

haukepetersen opened the pull request:
To get RIOT ready for the xtimer, here is the low-level timer implementation for all MSP430 based boards. For backward compatibility, all effected boards do now use the `hwtimer_compat` module.

Please note, that the `vtimer` will not work correctly anymore, as it depends on a 32-bit `hwtimer` (while we now only have 16-bit wide low-level timers). But this should not matter much, as the boards run perfectly with the `xtimer`!

Note2 (especially @kaspar030): So far there is almost identical code for the `cc430` and the `msp430fxyz`. Especially for some other peripherals, they seem to differ quite a bit, but I am not sure yet to which extend. So the strategy is clearly for now: implement separately and merge again as soon as I have a better feeling for those MCUs - so be patient :-)

PRs for other peripheral timers will follow soon...

kaspar030 commented:
Nice! will test ASAP.

(you probably need to (void) some arguments)

haukepetersen commented:
@kaspar030, any progress testing this?

kaspar030 commented:
nope, busy, will do later today!

kaspar030 commented:
Works as expected on msb-430h. Can't test on other boards. Do you want to fix the "N" define in xtimer or the msb headers? ACK after that is fixed.

haukepetersen commented:
Just fixing it as we speak..

haukepetersen commented:
fixed the xtimer tests that Travis was complaining about and removed the TI legacy header (`legacymsp430.h`). The xtimer test should work fine on all effected boards.

kaspar030 commented:
Tested again on msb-430h, ACK holds from my side if travis is happy.

OlegHahm commented:
Will test on Z1 and WSN430 now.

OlegHahm commented:
Hm, on Z1 with `example/timer_periodic_wakeup` I observe a weird pattern (look at the pyterm timestamp):
```
2015-09-02 17:31:06,363 - INFO # slept until 79065481
2015-09-02 17:31:07,348 - INFO # slept until 80065481
2015-09-02 17:31:08,335 - INFO # slept until 81065481
2015-09-02 17:31:08,338 - INFO # slept until 81002729
2015-09-02 17:31:08,341 - INFO # slept until 81005633
2015-09-02 17:31:08,344 - INFO # slept until 81008539
2015-09-02 17:31:08,346 - INFO # slept until 81011447
2015-09-02 17:31:08,349 - INFO # slept until 81014352
2015-09-02 17:31:08,352 - INFO # slept until 81017263
2015-09-02 17:31:08,355 - INFO # slept until 81020174
2015-09-02 17:31:08,358 - INFO # slept until 81023079
2015-09-02 17:31:08,361 - INFO # slept until 81025989
2015-09-02 17:31:08,364 - INFO # slept until 81028893
2015-09-02 17:31:08,367 - INFO # slept until 81031799
2015-09-02 17:31:08,369 - INFO # slept until 81034706
2015-09-02 17:31:08,372 - INFO # slept until 81037615
2015-09-02 17:31:08,375 - INFO # slept until 81040522
2015-09-02 17:31:08,378 - INFO # slept until 81043426
2015-09-02 17:31:08,381 - INFO # slept until 81046332
2015-09-02 17:31:08,384 - INFO # slept until 81049242
...
```
The node seems to run fine for some while and then trigger a lot of interrupts at once, go back to normal mode and so on...

haukepetersen commented:
I can reproduce this behavior on the `msb-430h`. @kaspar030, any hint if this is `xtimer` related or if is this caused by the peripheral timer?

OlegHahm commented:
For the record: very similar pattern on wsn430-v1_3b

kaspar030 commented:
same on msb430h. hm, whenever that happens, the actual sleep time goes backwards by around 63k us, which sounds like a 16bit handling issue.

kaspar030 commented:
@haukepetersen yes, let me try to rule out xtimer stuff, but it actually looks like that.

haukepetersen commented:
yap. Is there maybe some mask define for 16-bit timers that I missed to set for the boards, but xtimer needs?

OlegHahm commented:
I don't know if related and/or helpful, but `tests/xtimer_usleep_until` terminates at
```
2015-09-02 17:46:34,409 - INFO # Testing interval 95...
2015-09-02 17:46:34,415 - INFO # Testing interval 94...
2015-09-02 17:46:34,416 - INFO # Testing interval 93...
2015-09-02 17:46:34,416 - INFO # Testing interval 92...
2015-09-02 17:46:34,434 - INFO # Testing interval 91...
2015-09-02 17:46:34,435 - INFO # Testing interval 90...
2015-09-02 17:46:34,435 - INFO # Testing interval 89...
2015-09-02 17:46:34,436 - INFO # Testing interval 88...
2015-09-02 17:46:34,436 - INFO # Testing interval 87...
2015-09-02 17:46:34,437 - INFO # Testing interval 86...
2015-09-02 17:46:34,437 - INFO # Testing interval 85...
2015-09-02 17:46:34,437 - INFO # Testing interval 84...
2015-09-02 17:46:34,438 - INFO # Testing interval 83...
2015-09-02 17:46:34,439 - INFO # Testing interval 82...
2015-09-02 17:46:34,439 - INFO # Testing interval 81...
2015-09-02 17:46:34,505 - INFO # Testing interval 80...
2015-09-02 17:46:34,572 - INFO # Testing interval 79...
2015-09-02 17:46:34,639 - INFO # Testing interval 78...
2015-09-02 17:46:34,705 - INFO # Testing interval 77...
2015-09-02 17:46:34,772 - INFO # Testing interval 76...
2015-09-02 17:46:34,839 - INFO # Testing interval 75...
2015-09-02 17:46:34,905 - INFO # Testing interval 74...
2015-09-02 17:46:34,972 - INFO # Testing interval 73...
2015-09-02 17:46:35,039 - INFO # Testing interval 72...
2015-09-02 17:46:35,105 - INFO # Testing interval 71...
```

OlegHahm commented on boards/chronos/include/periph_conf.h:
```diff
@@ -14,17 +15,28 @@
-#ifndef PERIPH_CONF_H_
-#define PERIPH_CONF_H_
+#ifndef PERIPH_CONF_H
+#define PERIPH_CONF_H
```
minor: but why did you revert this change?

haukepetersen replied:
propably copy+paste. I tend to use guards without trailing underscores in the files I create...

OlegHahm replied:
No strong feelings, would just vote for consistency and currently we have 36 `PERIPH_CONF_H_`s and only 6 `PERIPH_CONF_H`s.

haukepetersen replied:
I don't care, I can offer you to open a follow-up PR to make them all consistent. Would that work?

OlegHahm replied:
works for me

haukepetersen commented:
re-introduced `eINT` and `dINT`, as it seems that the `pthread` implementations still use them and Travis fails without them...

kaspar030 commented:
> re-introduced eINT and dINT, as it seems that the pthread implementations still use them and Travis fails without them...

Uff, that should seriously be fixed in pthreads. Promise to at least create an issue. ;)

kaspar030 commented:
I have the feeling that this somehow loses interrupts, but I can't pinpoint it. I'm on it.

kaspar030 commented:
No, my problem is xtimer related. ACK when travis is happy!

haukepetersen commented:
nice :-)

haukepetersen commented:
Travis is happy -> and go!

kaspar030 commented:
yay!
